### ISSUE TITLE
Zelda OOT issue with fragment based depth

### ISSUE THREAD

**Issue by fzurita** [gonetz/GLideN64#2382]
With fragment based depth enabled I get this issue when zelda OOT at the very beginning of the game:
![the_legend_of_zelda-000](https://user-images.githubusercontent.com/11581687/98878484-1f3d6400-2451-11eb-811b-0475fecc7cd7.png)

Not sure when the problem was introduced. I tried to go back pretty far, but once CMake stops being used for android, it makes it really hard to compile.

This is in an Adreno GPU.

**Comment by fzurita** [comment]
Ok, this will only happen with Adreno GPUs because it's the only GLES GPU that also supports GL_NV_shader_noperspective_interpolation. What we are doing with this extension is causing this problem.

**Comment by gonetz** [comment]
As I see, GL_NV_shader_noperspective_interpolatio was introduced by Logan in 2e227cef531c1
It was April 2018.
@fzurita as I understand, it is hard for you to build the project from the sources that old. Can you revert that commit from the current master and check, is it really the source of the problem? I have a device with Adreno 420, but I can't reproduce that issue with it. Probably it does not support that extension.

**Comment by fzurita** [comment]
@gonetz if I force the extension to disabled, the problem goes away. Although, it would be good to find out if the problem existed when it was first introduced.

If you are using the Mupen64plus AE or FZ app, you should be able to reproduce it using the GLideN64-Very-Accurate profile.

**Comment by gonetz** [comment]
Yes, I use Mupen64plus FZ, ver 251. I can't reproduce that issue on my device with  GLideN64-Very-Accurate profile.



> Although, it would be good to find out if the problem existed when it was first introduced.

Yes. As I understand, it is enough to enable only "write fragment based depth" option to reproduce that error. N64 depth compare can be off. If so, I suppose these changes in glsl_CombinerProgramBuilder.cpp  can be responsible:

```
			ss << "# define IN in" << std::endl << "# define OUT out" << std::endl;
-			if (_glinfo.noPerspective)
-				ss << "noperspective OUT highp float vZCoord;" << std::endl << "uniform lowp int uClampMode;" << std::endl;
+			if (_glinfo.noPerspective) {
+				ss << "#extension GL_NV_shader_noperspective_interpolation : enable" << std::endl
+					<< "noperspective OUT highp float vZCoord;" << std::endl << "uniform lowp int uClampMode;" << std::endl;
+			}
			m_part = ss.str();
```

Probably noperspective in "noperspective OUT highp float vZCoord;" was ignored without  "#extension GL_NV_shader_noperspective_interpolation : enable". Could you check it? Other changes in glsl_CombinerProgramBuilder.cpp are related to N64 depth compare and should not affect "write fragment based depth" option.

**Comment by fzurita** [comment]
Reverting those few lines doesn't fix it. I'll try to see what happens if try to go back to the initial state when @loganmc10 first introduced this.

**Comment by fzurita** [comment]
Also, I don't have N64 depth compare enabled, so I'm not sure how it could be causing an issue. If I enable N64 depth compare, things get much worse, link goes completely invisible.

Also, I checked your GPU, it doesn't have that extension: https://opengles.gpuinfo.org/displayreport.php?id=4247

**Comment by fzurita** [comment]
I think it's more likely that it's a problem with GL depth clamp emulation in GLES:
https://github.com/gonetz/GLideN64/pull/1747
https://github.com/gonetz/GLideN64/pull/1783

**Comment by fzurita** [comment]
Yeah, I can confirm that disabling the cod where ```GL_NV_shader_noperspective_interpolation``` checks were added in this commit fixes the issue, but breaks the slingshot again:

https://github.com/gonetz/GLideN64/pull/1783

**Comment by fzurita** [comment]
Looking closer, it may be related to ```ShaderFragmentRenderTarget```. It also touches the fragment depth and it seems to be in use while the backgrounds are being rendered. 

I have the feeling that things are being scaled differently between that shader and ```ShaderWriteDepth``` when ```GL_NV_shader_noperspective_interpolation``` is present.

**Comment by gonetz** [comment]
> Yeah, I can confirm that disabling the code where `GL_NV_shader_noperspective_interpolation` checks were added in this commit fixes the issue, but breaks the slingshot again:

Yes, Logan explained in #1783 that noperspective is required for depth clamp emulation.

> Looking closer, it may be related to `ShaderFragmentRenderTarget`. It also touches the fragment depth and it seems to be in use while the backgrounds are being rendered.

`ShaderFragmentRenderTarget` is always in use. It writes correct value to gl_FragDepth, that is depth buffer value.
The depth buffer value is calculated in `ShaderWriteDepth` from gl_FragCoord.z (GL) or from vZCoord (GLES).
However, if current rendering buffer is depth buffer, as it happens during depth buffer clear, depth buffer value is taken from fragment color R component:
```
  if (uRenderTarget != 0) {
   ...
    fragDepth = fragColor.r;
  }

```


> have the feeling that things are being scaled differently between that shader and `ShaderWriteDepth` when `GL_NV_shader_noperspective_interpolation` is present.

Lets see:
```
if (_glinfo.isGLESX && _glinfo.noPerspective) {
        m_part =
       	"highp float writeDepth()									\n"
       	"{												\n"
       	"  if (uClampMode == 1 && (vZCoord > 1.0)) discard;						\n"
       	"  if (uDepthSource != 0) return uPrimDepth;							\n"
       	"  return clamp((vZCoord - uPolygonOffset) * uDepthScale.s + uDepthScale.t, 0.0, 1.0);		\n"
       	"}												\n"
       	;                                                                                               

```

`if (uClampMode == 1 && (vZCoord > 1.0)) discard;`  line emulates the depth clamp.

vZCoord is calculated in vertex shader as `vZCoord = gl_Position.z / gl_Position.w;` and passed as noperspective varying parameter to the fragment shader. GL code uses 
 `return clamp((gl_FragCoord.z * 2.0 - 1.0) * uDepthScale.s + uDepthScale.t, 0.0, 1.0);` instead.
Is `(gl_FragCoord.z * 2.0 - 1.0)` equivalent to `(vZCoord - uPolygonOffset)`? I'm not sure, but at least it works on NVidia Shields, does not it? Probably we have a precision issue here. May be a value of uPrimDepth somehow breaks the depth compare. You may try to hack that part:
1. Try     	`"  if (uDepthSource != 0) return 1.0;							\n"`
2. Try to remove 	`"  if (uClampMode == 1 && (vZCoord > 1.0)) discard;						\n"`

If one of these hacks helps to fix the issue, it may give as a hint for a general solution.
It is hard for me to guess blindly without ability to test it on real device.

**Comment by fzurita** [comment]
I tried the second one with no improvement yesterday. I'll try the first hack though.

Also, of you have a NVidia graphics card, it probably supports all the right extensions. You could force the GLES path for the NVidia graphics card to see what happens.

**Comment by standard-two-simplex** [comment]
>  `vZCoord = gl_Position.z / gl_Position.w;`

I had some issues when performing a similar division while investigating for the `use_screen_coords` branch.

Apparently, after muliplication by a perspective projection matrix, `w` can hold values which are negative, zero or positive but very close to zero. This means that after division by `w` this vertices can end behind the far plane, undefined or very far from the near plane respectively. To avoid this problem, OpenGL first clips the `w` coordinate to a certain internal threshold (I believe), then clips everything into the view frustrum and finally performs division by `w`, when `w` is guaranteed to be bigger than the threshold and the near plane `w`.

I had issues with performing such division in `processVertex()` and I can't see why it would be any different in the vertex shader. You may want to check if there's any geometry redered with vertex `w` lower or equal to 0. Maybe there's a ground polygon with a vertex behind the camera or the triangles (are they triangles?) that render the background.

I can't see any good solution for such issue besides clipping in software or leaving the coordinates unmodified (i.e. in clip space) and letting OpenGL handle the clipping.

**Comment by fzurita** [comment]
@gonetz I tested both hacks and neither solved the issue.
@standard-two-simplex I wonder if there is a way to apply the ```GL_NV_shader_noperspective_interpolation``` way of writing depth only when we need to disable near plane clipping. That way other cases use the normal logic which works normally.

**Comment by fzurita** [comment]
Ok, I'm making some progress. What is blocking link's view is the table in the middle of the room. In this case, it appears that near clippings plane need to be enabled so that the table doesn't show up in the depth buffer.

**Comment by fzurita** [comment]
Fixed 402cf6f8b6e5a105ced42010965a1d72ff3f7f43
